kong/clustering/wrpc_control_plane.lua

```diff
@@ -0,0 +1,662 @@
+
+      elseif not ok then
+        if self.conf.cluster_ocsp == "on" then
+          err = "data plane client certificate revocation check failed: " .. err
+
+        else
+          ngx_log(ngx_WARN, _log_prefix, "data plane client certificate revocation check failed: ", err, log_suffix)
+          err = nil
+        end
+      end
+    end
+
+    if err then
+      ngx_log(ngx_ERR, _log_prefix, err, log_suffix)
+      return ngx_exit(ngx_CLOSE)
+    end
+  end
+
+  if not dp_id then
+    ngx_log(ngx_WARN, _log_prefix, "data plane didn't pass the id", log_suffix)
+    ngx_exit(400)
+  end
+
+  if not dp_version then
+    ngx_log(ngx_WARN, _log_prefix, "data plane didn't pass the version", log_suffix)
+    ngx_exit(400)
+  end
+
+  local wb
+  do
+    local err
+    wb, err = ws_server:new(WS_OPTS)
+    if not wb then
+      ngx_log(ngx_ERR, _log_prefix, "failed to perform server side websocket handshake: ", err, log_suffix)
+      return ngx_exit(ngx_CLOSE)
+    end
+  end
+
+  -- connection established
+  local w_peer = wrpc.new_peer(wb, get_config_service(self))
+  local client = {
+    last_seen = ngx_time(),
+    peer = w_peer,
+    dp_id = dp_id,
+    dp_version = dp_version,
+    log_suffix = log_suffix,
+    basic_info = nil,
+    basic_info_semaphore = semaphore.new()
+  }
+  self.clients[w_peer.conn] = client
+  w_peer:spawn_threads()
+
+  do
+    local ok, err = client.basic_info_semaphore:wait(5)
+    if not ok then
+      err = "waiting for basic info call: " .. (err or "--")
+    end
+    if not client.basic_info then
+      err = "invalid basic_info data"
+    end
+
+    if err then
+      ngx_log(ngx_ERR, _log_prefix, err, log_suffix)
+      wb:send_close()
+      return ngx_exit(ngx_CLOSE)
+    end
+  end
+
+  client.dp_plugins_map = plugins_list_to_map(client.basic_info.plugins)
+  client.config_hash = string.rep("0", 32) -- initial hash
+  client.sync_status = CLUSTERING_SYNC_STATUS.UNKNOWN
+  local purge_delay = self.conf.cluster_data_plane_purge_delay
+  function client:update_sync_status()
+    local ok, err = kong.db.clustering_data_planes:upsert({ id = dp_id, }, {
+      last_seen = self.last_seen,
+      config_hash = self.config_hash ~= "" and self.config_hash or nil,
+      hostname = dp_hostname,
+      ip = dp_ip,
+      version = dp_version,
+      sync_status = self.sync_status, -- TODO: import may have been failed though
+    }, { ttl = purge_delay })
+    if not ok then
+      ngx_log(ngx_ERR, _log_prefix, "unable to update clustering data plane status: ", err, log_suffix)
+    end
+  end
+
+  do
+    local _, err
+    _, err, client.sync_status = self:check_version_compatibility(dp_version, client.dp_plugins_map, log_suffix)
+    if err then
+      ngx_log(ngx_ERR, _log_prefix, err, log_suffix)
+      wb:send_close()
+      client:update_sync_status()
+      return ngx_exit(ngx_CLOSE)
+    end
+  end
+
+  self:push_config()    -- first config push
```

Comment:
Won't this cause a config push to all nodes for every new connection?
Maybe this would do.

```suggestion
  do
    local payload, err = self:export_deflated_reconfigure_payload()
    if not payload then
      ngx_log(ngx_ERR, _log_prefix, "unable to export config from database: ", err)
      return
    end
    ngx_log(ngx_NOTICE, _log_prefix, "config version #", config_version, " pushed to ", dp_id)
  end
```

Reply:
Also, maybe we can cache the "payload" to speed up the case, where new nodes come and the config is not modified.

Reply:
good call.

Reply:
done!